lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -322,6 +322,104 @@ void SPIRVRegularizeLLVMBase::expandSYCLTypeUsing(Module *M) {
+// Rule 1: If the argument is any NaN, zero is returned.
+// Rule 2: If the argument is smaller than the smallest representable
+// (un)signed integer of the result type, the smallest representable
+// (un)signed integer is returned.
+// Rule 3: If the argument is larger than the largest representable (un)signed
+// integer of the result type, the largest representable (un)signed integer is
+// returned.
+// Rule 4: Otherwise, the result of rounding the argument towards zero is
+// returned.
+// Rules 1 & 4 are preserved when promoting iX to i64. For preserving Rule 2
+// and Rule 3, we saturate the result of the promoted instruction based on
+// original integer type (iX)
+// Example:
+// Input:
+// %0 = call i2 @llvm.fptosi.sat.i2.f32(float %input)
+// %1 = sext i32 %0
+// Output:
+// %0 = call i32 @_Z17convert_long_satf(float %input)
+// %1 = icmp sge i32 %0, 1 <Largest 2-bit signed integer>
+// %2 = icmp sle i32 %0, -2 <Smallest 2-bit signed integer>
+// %3 = select i1 %1, i32 1, i32 %0
+// %4 = select i1 %2, i32 -2, i32 %3
+// Replace uses of %1 in Input with %4 in Output
+void SPIRVRegularizeLLVMBase::cleanupConversionToNonStdIntegers(Module *M) {
+  for (auto I = M->begin(), E = M->end(); I != E;) {
+    Function *F = &(*I++);
+    std::vector<Instruction *> ToErase;
+    for (BasicBlock &BB : *F) {
+      for (Instruction &I : BB) {
+        if (IntrinsicInst *II = dyn_cast<IntrinsicInst>(&I)) {
+          // TODO: Vector type not supported yet.
+          if (isa<VectorType>(II->getType()))
+            continue;
+          auto IID = II->getIntrinsicID();
+          auto IntBitWidth = II->getType()->getScalarSizeInBits();
+          if (IntBitWidth == 8 || IntBitWidth == 16 || IntBitWidth == 32 ||
+              IntBitWidth == 64)
+            continue;
+          if (IID == Intrinsic::fptosi_sat) {
+            // Identify sext (user of I). Make sure that's the only use of I.
+            auto *User = I.getUniqueUndroppableUser();
+            if (!User || !isa<SExtInst>(User))
+              continue;
+            auto *SExtI = dyn_cast<SExtInst>(User);
+            auto *NewIType = SExtI->getType();
+            IRBuilder<> IRB(&I);
+            auto *NewII = IRB.CreateIntrinsic(
+                IID, {NewIType, II->getOperand(0)->getType()},
+                II->getOperand(0));
+            Constant *MaxVal = ConstantInt::get(
+                NewIType, APInt::getSignedMaxValue(IntBitWidth).getSExtValue());
+            Constant *MinVal = ConstantInt::get(
+                NewIType, APInt::getSignedMinValue(IntBitWidth).getSExtValue());
+            auto *GTMax = IRB.CreateICmp(CmpInst::ICMP_SGE, NewII, MaxVal);
+            auto *LTMin = IRB.CreateICmp(CmpInst::ICMP_SLE, NewII, MinVal);
+            auto *SatMax = IRB.CreateSelect(GTMax, MaxVal, NewII);
+            auto *SatMin = IRB.CreateSelect(LTMin, MinVal, SatMax);
+            SExtI->replaceAllUsesWith(SatMin);
```

Comment:
Moving this under: `for (Instruction *V : ToErase) {` would allow multiple-users case

Reply:
We do not expect multiple users here.